lib/model/indexhandler.go

```diff
@@ -222,15 +222,35 @@ func (s *indexHandler) pause() {
 // sendIndexTo sends file infos with a sequence number higher than prevSequence and
 // returns the highest sent sequence number.
 func (s *indexHandler) sendIndexTo(ctx context.Context, fset *db.FileSet) error {
+	// Keep track of the previous sequence we sent. This is separate from
+	// s.prevSequence because the latter will skip over holes in the
+	// sequence numberings, while sentPrevSequence should always be
+	// precisely the highest previousely sent sequence.
+	sentPrevSequence := s.prevSequence
+
 	initial := s.prevSequence == 0
 	batch := db.NewFileInfoBatch(nil)
 	batch.SetFlushFunc(func(fs []protocol.FileInfo) error {
+		var lastSequence int64
+		if len(fs) > 0 {
+			lastSequence = fs[len(fs)-1].Sequence
+		}
 		l.Debugf("%v: Sending %d files (<%d bytes)", s, len(fs), batch.Size())
 		if initial {
 			initial = false
-			return s.conn.Index(ctx, &protocol.Index{Folder: s.folder, Files: fs})
+			return s.conn.Index(ctx, &protocol.Index{
+				Folder:       s.folder,
+				Files:        fs,
+				LastSequence: lastSequence,
+			})
 		}
-		return s.conn.IndexUpdate(ctx, &protocol.IndexUpdate{Folder: s.folder, Files: fs})
+		defer func() { sentPrevSequence = lastSequence }()
```

Comment:
That's valid, but I think can be left off at the moment. In principle the batcher could/should panic with a bug message if the flush function is called again after an error is returned, as given how we're using it that's clearly unsafe as we don't know the status of the previous flush...

Reply:
Generally fully agree that it's not strictly necessary right now.  
The batch reuse on error concern seems to be orthogonal to my concern: Even if the batch is reset/otherwise not reused, `sentPrevSequence` is not tied to that. If on error one resets the batch and retry, one needs `sentPrevSequence` to be at the value before the send attempt that produced the error.